### DISCUSSION TITLE
Multiple Algolia search index?

### DISCUSSION THREAD

**Discussion by baptisteArno** [facebook/docusaurus#8467] [Q&A]
I have a doc with a sub path `/api` that points to the API doc.

I'd like to change the Algolia search index based on this:

If user in documentation (every page other than `/api/**`) then don't show search results from Typebot API index
Otherwise, use the API index

<img width="896" alt="CleanShot 2022-12-20 at 10 32 01@2x" src="https://user-images.githubusercontent.com/16015833/208633121-a93f2b1e-540f-4dda-83c2-738ebe4656aa.png">

How can I achieve this?

**Comment by slorber** [comment]
Hi

We assign metadata tags to each page for the search crawler.

For example your regular docs: https://docs.typebot.io/

```
<meta name="docsearch:docusaurus_tag" content="docs-default-current" data-rh="true">
```

Your API ref docs: https://docs.typebot.io/api

```html
<meta name="docusaurus_tag" content="default" data-rh="true">
```

The plugin you use for the API docs does not populate `docusaurus_tag` to make it easy to index/query those pages differently from the crawler, but you can. probably swizzle one of its theme components to add this tag yourself with our `<Head>` API


Now we have the concept of "contextual facet filters" which permit to craft an Algolia query based on the current path.


```js
// Translate search-engine agnostic search filters to Algolia search filters
export function useAlgoliaContextualFacetFilters(): [string, string[]] {
  const {locale, tags} = useContextualSearchFilters();

  // Seems safe to convert locale->language, see AlgoliaSearchMetadata comment
  const languageFilter = `language:${locale}`;

  const tagsFilter = tags.map((tag) => `docusaurus_tag:${tag}`);
  
  // [] = AND query
  return [languageFilter, tagsFilter];
}
```

Unfortunately I don't think you can swizzle that file anymore but you can probably create a Webpack alias to override it with your own implementation, and use the `useLocation` hook to do different things based on the current path

```js
// Translate search-engine agnostic search filters to Algolia search filters
export function useAlgoliaContextualFacetFilters(): [string, string[]] {
  const {pathname} = useLocation()
  const {locale, tags} = useContextualSearchFilters();

  if (pathname.startsWith("/api/") { 
    return ["docusaurus_tag:my-api-ref-tag"]
  }

  // Seems safe to convert locale->language, see AlgoliaSearchMetadata comment
  const languageFilter = `language:${locale}`;

  const tagsFilter = tags.map((tag) => `docusaurus_tag:${tag}`);
  
  // [] = AND query
  return [languageFilter, tagsFilter];
}
```

---


Not sure we provide any better way to do that currently.

Maybe @shortcuts has another idea?

**Reply by shortcuts** [reply]
Hey! For simplicity I'd avise you to do a single index with filters, since it works out of the box for both DocSearch and Docusaurus.

My idea is similar, but it's also possible to do the filtering part on the Algolia side. Inside your Crawler configuration, you can leverage the [`url` parameter](https://www.algolia.com/doc/tools/crawler/guides/extracting-data/#anatomy-of-a-recordextractor) of the `recordExtractor` to create a facet based on the path:
```js
      recordExtractor: ({ $, helpers, url }) => {
        const filter = url.pathname.includes("...") ? "a" : "b"

        // ... rest of the logic
      }
```

and then follow https://docsearch.algolia.com/docs/record-extractor/#indexing-content-for-faceting

You'd still however to make it dynamic on the contextual search for Docusaurus

**Reply by baptisteArno** [reply]
Thank you so much for the info! 🙏

https://docs.typebot.io/ has `<meta name="docsearch:docusaurus_tag" content="docs-default-current" data-rh="true">`
And https://docs.typebot.io/api has `<meta name="docsearch:docusaurus_tag" content="default" data-rh="true">`

Contextual search should work properly since `docsearch:docusaurus_tag` content is different? `"docs-default-current" !== "default"`

What am I missing?

**Reply by baptisteArno** [reply]
Here is my crawler config:

```js
new Crawler({
  rateLimit: 8,
  startUrls: ["https://docs.typebot.io/"],
  renderJavaScript: false,
  sitemaps: ["https://docs.typebot.io//sitemap.xml"],
  ignoreCanonicalTo: true,
  discoveryPatterns: ["https://docs.typebot.io/**"],
  exclusionPatterns: ["https://docs.typebot.io/self-hosting/configuration/**"],
  schedule: "at 3:19 PM on Monday",
  actions: [
    {
      indexName: "typebot",
      pathsToMatch: ["https://docs.typebot.io/**"],
      recordExtractor: ({ $, helpers }) => {
        // priority order: deepest active sub list header -> navbar active item -> 'Documentation'
        const lvl0 =
          $(
            ".menu__link.menu__link--sublist.menu__link--active, .navbar__item.navbar__link--active"
          )
            .last()
            .text() || "Documentation";
        return helpers.docsearch({
          recordProps: {
            lvl0: {
              selectors: "",
              defaultValue: lvl0,
            },
            lvl1: ["header h1", "article h1"],
            lvl2: "article h2",
            lvl3: "article h3",
            lvl4: "article h4",
            lvl5: "article h5, article td:first-child",
            lvl6: "article h6",
            content: "article p, article li, article td:last-child",
          },
          aggregateContent: true,
          recordVersion: "v3",
        });
      },
    },
  ],
  initialIndexSettings: {
    typebot: {
      attributesForFaceting: [
        "type",
        "lang",
        "language",
        "version",
        "docusaurus_tag",
      ],
      attributesToRetrieve: [
        "hierarchy",
        "content",
        "anchor",
        "url",
        "url_without_anchor",
        "type",
      ],
      attributesToHighlight: ["hierarchy", "hierarchy_camel", "content"],
      attributesToSnippet: ["content:10"],
      camelCaseAttributes: ["hierarchy", "hierarchy_radio", "content"],
      searchableAttributes: [
        "unordered(hierarchy_radio_camel.lvl0)",
        "unordered(hierarchy_radio.lvl0)",
        "unordered(hierarchy_radio_camel.lvl1)",
        "unordered(hierarchy_radio.lvl1)",
        "unordered(hierarchy_radio_camel.lvl2)",
        "unordered(hierarchy_radio.lvl2)",
        "unordered(hierarchy_radio_camel.lvl3)",
        "unordered(hierarchy_radio.lvl3)",
        "unordered(hierarchy_radio_camel.lvl4)",
        "unordered(hierarchy_radio.lvl4)",
        "unordered(hierarchy_radio_camel.lvl5)",
        "unordered(hierarchy_radio.lvl5)",
        "unordered(hierarchy_radio_camel.lvl6)",
        "unordered(hierarchy_radio.lvl6)",
        "unordered(hierarchy_camel.lvl0)",
        "unordered(hierarchy.lvl0)",
        "unordered(hierarchy_camel.lvl1)",
        "unordered(hierarchy.lvl1)",
        "unordered(hierarchy_camel.lvl2)",
        "unordered(hierarchy.lvl2)",
        "unordered(hierarchy_camel.lvl3)",
        "unordered(hierarchy.lvl3)",
        "unordered(hierarchy_camel.lvl4)",
        "unordered(hierarchy.lvl4)",
        "unordered(hierarchy_camel.lvl5)",
        "unordered(hierarchy.lvl5)",
        "unordered(hierarchy_camel.lvl6)",
        "unordered(hierarchy.lvl6)",
        "content",
      ],
      distinct: true,
      attributeForDistinct: "url",
      customRanking: [
        "desc(weight.pageRank)",
        "desc(weight.level)",
        "asc(weight.position)",
      ],
      ranking: [
        "words",
        "filters",
        "typo",
        "attribute",
        "proximity",
        "exact",
        "custom",
      ],
      highlightPreTag: '<span class="algolia-docsearch-suggestion--highlight">',
      highlightPostTag: "</span>",
      minWordSizefor1Typo: 3,
      minWordSizefor2Typos: 7,
      allowTyposOnNumericTokens: false,
      minProximity: 1,
      ignorePlurals: true,
      advancedSyntax: true,
      attributeCriteriaComputedByMinProximity: true,
      removeWordsIfNoResults: "allOptional",
    },
  },
  appId: "...",
  apiKey: "...",
  indexPrefix: "",
});
```

**Reply by baptisteArno** [reply]
Is this a bug? Should I fill an issue? @slorber

**Reply by slorber** [reply]
First, I'm surprised that your API section has `<meta name="docusaurus_tag" content="default" data-rh="true">`, it shouldn't have "default" but "docs-default-current" or "docs-api-current" or something similar. 

It is likely that you changed something that lead to this default value, or do not use the latest Docusaurus version or whatever but it's not normal. If I can't see your code/config, I can't help.

**Edit**: oh this is because you use a third-party OpenAPI plugin. Each plugin is responsible to declare the docusaurus search tag it wants for each individual page.

---

Second, Contextual search is meant to apply some dynamic filters for you automatically. 
If you are on the regular docs, it is not meant to prevent displaying search results from your API or landing page.

The main goals are:
- If you are on the FR site, you should only display the FR search results
- If you are on API v1, you should only display results of API v1 (and not duplicates from API v1 + v2 + v3 + v4...)

Imagine a site where you have iOS v1 + v2 (current), and Android v1 + v2 (current).
Browsing the iOS docs in v1 and triggering a search with contextual search enabled doesn't mean you won't display Android search results.

It means you will search in any of:
- iOS v1 docs (**and not v2 versions** because you are browsing iOS v1)
- Android v2 (because it is the current version)
- All the blog pages
- All the landing pages like your home page

Contextual search is not a way to "sandbox" the search to a single part of your site but rather a way to avoid displaying duplicate search results due to localization or versioning. Contextual search always search inside all pages that have the "default" docusaurus_tag, that is set by your OpenAPI plugin.
